statedb: Add Kubernetes StateDB reflector

joamaki opened the pull request:
Adds the reflector package and the `Kubernetes` reflector which essentially takes a `cache.ListerWatcher` and writes the changing objects into the given table.

Also adds `statedb.Observable[Obj any](*DB, Table[Obj) stream.Observable[Event[Obj]]` for `Resource[T]` style stream of events to make it easier to migrate from `Resource[T]` to `Table[T]`. Won't support workqueue for retries, for that we'd implement the workqueue directly and perhaps use the stream to queue the work and then query the table (though when retrying is needed it's better to use the generic reconciler).

joamaki committed:
stream: Add Buffer operator

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb/reflector: Add Kubernetes reflector

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/sources/common.go:
```diff
@@ -0,0 +1,7 @@
+
+import "github.com/cilium/cilium/pkg/hive"
+
+type Source[Obj any] interface {
```
hmm. `Origin`? `Creator`? `Ingress`? `Ingester`? `Producer`? `Generator`?

joamaki replied:
`Reflector` is what client-go used, so that's one option as well.

joamaki committed:
statedb: Add Observable function

As we already have lots of code doing processing via event streams (e.g. Resource[T]),
make it easier to migrate to the Kubernetes source from Resource[T] by making it possible
to observe the table as a stream.Observable. 

The workqueue that Resource[T] is not supported. It wasn't used in many cases anyway and
for those the workqueue can be implemented directly.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

pippolo84 commented on pkg/statedb/reflector/k8s.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+// Kubernetes synchronizes statedb table with an external Kubernetes resource.
+// Returns a 'Source' for starting and stopping it. If the source can be started
+// unconditionally, consider using [KubernetesCell] instead.
```
s/`'Source'`/`'Reflector'`

pippolo84 commented on pkg/statedb/reflector/k8s.go:
```diff
@@ -0,0 +1,254 @@
+					key = meta.GetName()
+				}
+
+				var ok bool
+				entry.obj, ok = transform(ev.Obj)
+				if !ok {
+					return buf
+				}
```
Nit: we can move this outside of the `if-else` since it is executed in both branches.

pippolo84 reviewed (commented):
LGTM
